### ISSUE TITLE
Question - Backend Service Integration

### ISSUE THREAD

**Issue by nithyanantham** [space10-community/conversational-form#71]
Thanks for bringing up new revolutionary web form. As part of some POC We are trying to integrate rocket.chat with conversational form or We want to control the bot flow based on backend services response maybe using NodeJS rest service or some other service.

Please let me know if there is any feature already exist or if there is any road map to implement such functionalities.

**Comment by felixnielsen** [comment]
@nithyanantham great you find interesting in Conversational Form!
If I understand correctly, then you want to be able to add a Webservice layer in-between questions and inject new tags (rendered from the server)?

**Comment by nithyanantham** [comment]
Yes that's correct

**Comment by felixnielsen** [comment]
@nithyanantham the current version does not support this "out of the box". It would be ok easy to implement. All the logic is there. It would be a combination between exposing a method to add tags at runtime, atm. tags are added initially.
If you want to give it a go yourself, then looking at how the "auto" tag [detection is done ](https://github.com/space10-community/conversational-form/blob/master/src/scripts/cf/ConversationalForm.ts#L170)
Together with https://github.com/space10-community/conversational-form#remaptagsandstartfrom you would get 95% of the way.
The feature itself is something we have in the pipeline, will be released in the coming weeks.

**Comment by felixnielsen** [comment]
@nithyanantham let me know if the above helped.

**Comment by sull** [comment]
Would this essentially be dynamic forms on-the-fly based on responses and CF would adapt?

**Comment by felixnielsen** [comment]
@sull yes! We are working on a formless Conversational Form as we speak. Will be shipped this week. @nithyanantham fyi

**Comment by sull** [comment]
Very cool. Could you summarize the "formless" feature according to how it has been designed and developed? Look forward to checking it out.

**Comment by felixnielsen** [comment]
@sull sure, formless is a feature that allow us to build the Conversational Form without an actual form element in the DOM.

So your input is instead a JSON object (which is 1:1 with html tag attributes):
````JSON
tags:{
	{
		"tag": "input",
		"type": "text",
		"pattern": ".{5,10}",
		"cf-input-placeholder": "Input w. pattern attribute",
		"cf-error": "No less than 5 and no more than 10 characters",
		"cf-questions": "Keep between 5-10 characters"
	},
	...
}
````
This allow us to build a tool that can build a Conversational Form. Also, we will add logic that allows you to inject tags at "runtime", ex. communicating with a web-service that returns above JSON API and then inject new tags.
You are also able to get a form element from the Conversational Form formless instance, that you can append to your document. And all current features works as is.

**Comment by felixnielsen** [comment]
@sull @nithyanantham please see latest code on develop
[examples/formless.html](https://github.com/space10-community/conversational-form/blob/develop/examples/formless.html#L314) <- injects new tags at runtime

uses [JSON file](https://github.com/space10-community/conversational-form/blob/develop/examples/files/formless.json) to init Conversational Form without a form tag

**Comment by sull** [comment]
interesting. i'll take a deeper look today.

**Comment by sull** [comment]
@felixnielsen I am testing now. Actually using the conditional-flow branch. 
I'm about to generate a custom dynamic json to build a form from a backend system I had previously setup to do this exact task. The initial test works, though has some lag after each user response (attached video).

[cf-formless-lag.mp4.zip](https://github.com/space10-community/conversational-form/files/922920/cf-formless-lag.mp4.zip)

**Comment by felixnielsen** [comment]
hi @sull - the “lag” is that not caused by your API calls? I see that dev.php is going from status “pending" to “200”, ex. from second 5 to 8 in your video, taking approx 2-3 seconds to do the API call to **dev.php**

**Comment by felixnielsen** [comment]
Closing, API integration is possible from [v0.9.3](https://github.com/space10-community/conversational-form/releases/tag/v0.9.3) and @sull the lag seems to be related to the dev.php call, let me know.

**Comment by sull** [comment]
Yeah probably, sorry for assuming it was the CF code.
